### ISSUE TITLE
Add Range ResourceMapper

### ISSUE THREAD

**Issue by zhxnlai** [cashapp/misk#1336]
**tl;dr:** consistent hashing is the only way to load balance leases in misk. Although it is a well established way to partition databases, counter intuitively, it is not the best way to load balance event consumers, where the number of total leases is small and each consumer is stateless.

Our team operates a service that consumes an event topic. This topic has 32 partitions and emits 400+ events per second. At some time last week, the service was auto scaled 20 pods. 

Most pods were doing work. But according to the metrics, the load on these pods was not evenly distributed: 4 pods were idle, not consuming any events, while 2 pods were busy, consuming a lot of events.

After doing some math, I found this:
- Assuming the consistent hash ring assigns partitions to a specific pod (let’s call it pod A) 1 time out of 20 (given 20 pods)
- Given 32 partitions, we have the following binomial distribution:

<img width="612" alt="Screen Shot 2019-12-12 at 9 58 22 PM" src="https://user-images.githubusercontent.com/4694896/71133199-ae44ec00-21ae-11ea-8e89-01b8d3d3878d.png">

- the probability that pod A receives 0 partition is 0.19371,
- the probability that pod A receives 1 partitions is 0.32625
- the probability that pod A receives 2 partitions is 0.26615
- the probability that pod A receives 3 partitions is 0.14008
- the probability that pod A receives 4 partitions is 0.05345

This distribution, consistent with the production metrics, is not ideal: 4 pods are idle (consuming 0 partition), 4 pods take twice the load (consuming 3 - 4 partitions). 

As we increase the number of partitions, the distribution becomes more uniform:

**# of partitions = 10000**
- <img width="519" alt="Screen Shot 2019-12-12 at 10 22 32 PM" src="https://user-images.githubusercontent.com/4694896/71133200-ae44ec00-21ae-11ea-893e-52b6aa6454e1.png">

**# of partitions = 100000**
- <img width="522" alt="Screen Shot 2019-12-12 at 10 23 19 PM" src="https://user-images.githubusercontent.com/4694896/71133201-ae44ec00-21ae-11ea-8b58-ade23ba06548.png">

It is common for databases to have more than 1^6 rows. But it is not practical for a topic to have this many partitions.

There are other ways to assign partitions to worker pods. For example,
- kafka client provides a RoundRobinAssignor https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/RoundRobinAssignor.java and a RangeAssignor https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java#L68
- kinesis client uses something similar to RangeAssignor https://github.com/awslabs/amazon-kinesis-client/blob/master/amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseTaker.java

They yield more consistent and uniform results. We should provide similar options for stateless lease holders.

**Comment by mightyguava** [comment]
Can you add a checklist the steps you'll take to implement this?
